test/cri-metrics.bats

```diff
@@ -153,6 +160,69 @@ EOF
 		cgroup_memory_failcnt=$(cat "$CTR_CGROUP"/memory.failcnt) &&
 		metrics_memory_failcnt=$(echo "$metrics" | jq '.podMetrics[0].containerMetrics[0].metrics[] | select(.name == "container_memory_failcnt") | .value.value | tonumber')
 	[[ $metrics_memory_failcnt == "$cgroup_memory_failcnt" ]]
+}
+
+@test "container hugetlb metrics" {
+	CONTAINER_ENABLE_METRICS="true" CONTAINER_METRICS_PORT=$(free_port) setup_crio
+	cat << EOF > "$CRIO_CONFIG"
+[crio.stats]
+collection_period = 0
+included_pod_metrics = [
+    "network",
+    "cpu",
+    "hugetlb",
+    "memory",
+    "oom",
+]
+EOF
+	start_crio_no_setup
+	check_images
 
-	stop_crio
+	metrics_setup
+	set_container_pod_cgroup_root "" "$CONTAINER_ID"
+
+	# allocate static huge pages
+	old_pages=$(cat /proc/sys/vm/nr_hugepages)
+	if [[ $old_pages == "0" ]]; then
+		echo 1 | tee /proc/sys/vm/nr_hugepages
+	fi
+
+	# make use of the huge page in the container
+	crictl exec --sync "$CONTAINER_ID" /usr/bin/cp /testdata/usehugetlb.c /
+	crictl exec --sync "$CONTAINER_ID" /usr/bin/gcc /usehugetlb.c -o /usr/bin/usehugetlb
+	crictl exec --sync "$CONTAINER_ID" /usr/bin/usehugetlb &
+	usehugetlb_pid=$!
+
+	# wait a bit for metrics sync - tests are more flaky without this
+	sleep 1
+	metrics=$(crictl metricsp | jq '.podMetrics[0].containerMetrics[0].metrics[]')
+
+	# assert container_hugetlb_usage_bytes{pagesize="2MB"} == 2MB
+	metrics_hugetlb_usage_2mb=$(echo "$metrics" | jq 'select(.name == "container_hugetlb_usage_bytes" and any(.labelValues[]; . == "2MB")) | .value.value | tonumber')
+	[[ $metrics_hugetlb_usage_2mb == "2097152" ]]
+
+	# assert container_hugetlb_max_usage_bytes{pagesize="2MB"} == 0(cgroup v2) or 2MB(cgroup v1)
+	# cgroup v2 does not support hugetlb max usage stats
+	metrics_hugetlb_max_usage_2mb=$(echo "$metrics" | jq 'select(.name == "container_hugetlb_max_usage_bytes" and any(.labelValues[]; . == "2MB")) | .value.value | tonumber')
+	if is_cgroup_v2; then
+		cgroup_hugetlb_max_usage_2mb=0
+	else
+		cgroup_hugetlb_max_usage_2mb=2097152
+	fi
+	[[ $metrics_hugetlb_max_usage_2mb == "$cgroup_hugetlb_max_usage_2mb" ]]
+
+	# assert container_hugetlb_usage_bytes{pagesize="1GB"} == 0
+	metrics_hugetlb_usage_1gb=$(echo "$metrics" | jq 'select(.name == "container_hugetlb_usage_bytes" and any(.labelValues[]; . == "1GB")) | .value.value | tonumber')
+	[[ $metrics_hugetlb_usage_1gb == "0" ]]
+
+	# assert container_hugetlb_max_usage_bytes{pagesize="1GB"} == 0
+	metrics_hugetlb_max_usage_1gb=$(echo "$metrics" | jq 'select(.name == "container_hugetlb_max_usage_bytes" and any(.labelValues[]; . == "1GB")) | .value.value | tonumber')
+	[[ $metrics_hugetlb_max_usage_1gb == "0" ]]
+
+	# cleanup
+	kill $usehugetlb_pid || true
+	wait $usehugetlb_pid || true
+	if [[ $old_pages == "0" ]]; then
+		echo 0 | tee /proc/sys/vm/nr_hugepages
+	fi
```

Comment:
Can you make sure this always runs even if the test failed somewhere?
You can use teardown or https://bats-core.readthedocs.io/en/stable/writing-tests.html#bats-on-failure-hook